xarray/backends/h5netcdf_.py

```diff
@@ -113,6 +115,14 @@ def open_store_variable(self, name, var):
             # save source so __repr__ can detect if it's local or not
             encoding['source'] = self._filename
             encoding['original_shape'] = var.shape
+            vlen_dtype = h5py.check_dtype(vlen=var.dtype)
+            if vlen_dtype is not None:
+                if vlen_dtype is not unicode_type:  # pragma: no cover
+                    raise NotImplementedError('unexpected vlen dtype: {!r}'
```

Comment:
h5py uses special NumPy dtypes (`np.object_`  with `metadata`) for defining variable length types. It doesn't use `var.dtype.kind == 'U'`, because HDF5 doesn't have a unicode type with a memory model matching `np.unicode_`. (HDF5 implements UTF-8 with a fixed number of bytes for storage, not a fixed number of unicode characters like NumPy.)